File: packages/npm/lib/sync.js
import chalk from 'chalk'
import ora from 'ora'
import { password } from '@inquirer/prompts'
import { existsSync, readFileSync, mkdirSync, copyFileSync } from 'fs'
import { resolve, join } from 'path'
import { simpleGit } from 'simple-git'
import { tmpdir } from 'os'
import { mkdtempSync, rmSync } from 'fs'
import Ajv from 'ajv'
import addFormats from 'ajv-formats'
import { fileURLToPath } from 'url'
import { dirname } from 'path'
import { getToken, setToken } from './config.js'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

function loadProjectConfig(cwd) {
  const configPath = join(cwd, '.workbench')
  if (!existsSync(configPath)) return null
  try {
    return JSON.parse(readFileSync(configPath, 'utf8'))
  } catch {
    return null
  }
}

function loadSchema() {
  const candidates = [
    resolve(__dirname, '..', '..', '..', 'protocol', 'workbench.schema.json'),
    join(process.env.HOME || process.env.USERPROFILE, '.workbench', 'src', 'protocol', 'workbench.schema.json'),
  ]
  for (const p of candidates) {
    if (existsSync(p)) {
      try { return JSON.parse(readFileSync(p, 'utf8')) } catch { /* skip */ }
    }
  }
  return null
}

function friendlyGitError(err, phase) {
  const msg = err?.message || String(err)

  if (msg.includes('Authentication failed') || msg.includes('403') || msg.includes('401')) {
    return {
      what: `authentication denied — token may be expired or lack repo scope`,
      next: `Run: workbench sync (you will be prompted for a new token)`,
    }
  }
  if (msg.includes('not found') || msg.includes('404') || msg.includes('Repository not found')) {
    return {
      what: `repository not found`,
      next: `Verify the bridge repo name is correct and the token has access`,
    }
  }
  if (msg.includes('could not resolve host') || msg.includes('unable to access')) {
    return {
      what: `network error — could not reach GitHub`,
      next: `Check your internet connection and try again`,
    }
  }
  if (msg.includes('non-fast-forward') || msg.includes('rejected')) {
    return {
      what: `remote has newer commits than local`,
      next: `Pull the bridge repo manually and retry`,
    }
  }
  // Generic fallback — strip any token from error message
  const cleaned = msg.replace(/https:\/\/[^@\s]+@/g, 'https://***@')
  return {
    what: cleaned,
    next: `Check git configuration and retry`,
  }
}

function printGitError(phase, err) {
  const { what, next } = friendlyGitError(err, phase)
  console.error(chalk.red(`  ✗ git ${phase} failed: ${what}`))
  console.error(chalk.gray(`    → ${next}`))
}

async function resolveToken() {
  let token = getToken()
  if (token) return token

  console.log(chalk.yellow('  No GitHub token found in ~/.contextkeeperrc'))
  console.log(chalk.gray('  Create a token at: https://github.com/settings/tokens'))
  console.log(chalk.gray('  Required scope: repo (full control of private repositories)\n'))

  token = await password({
    message: 'GitHub Personal Access Token:',
    mask: '*',
  })

  if (!token) {
    console.error(chalk.red('  ✗ token prompt failed: no token provided'))
    console.error(chalk.gray('    → Create a token at https://github.com/settings/tokens and retry'))
    process.exit(1)
  }

  setToken(token)
  console.log(chalk.green('  Token saved to ~/.contextkeeperrc\n'))
  return token
}

function buildAuthUrl(bridgeRepo, token) {
  return `https://${token}@github.com/${bridgeRepo}.git`
}

export async function syncProject(options) {
  const cwd = resolve('.')
  const config = loadProjectConfig(cwd)

  console.log(chalk.cyan('\n  workbench sync\n'))

  if (!config && !options.bridge) {
    console.error(chalk.red('  ✗ sync failed: no .workbench config found'))
    console.error(chalk.gray('    → Run: workbench init, or pass --bridge <owner/repo>'))
    process.exit(1)
  }

  const bridgeRepo = options.bridge || config?.bridge_repo
  const projectName = config?.project_name || ''
  const stateVectorRel = config?.state_vector_path || 'handoff/STATE_VECTOR.json'
  const handoffRel = config?.handoff_path || 'docs/HANDOFF.md'

  if (!bridgeRepo) {
    console.error(chalk.red('  ✗ sync failed: no bridge repo configured'))
    console.error(chalk.gray('    → Run: workbench init, or pass --bridge <owner/repo>'))
    process.exit(1)
  }

  if (!projectName) {
    console.error(chalk.red('  ✗ sync failed: no project_name in .workbench config'))
    console.error(chalk.gray('    → Run: workbench init to regenerate the config'))
    process.exit(1)
  }

  // Validate STATE_VECTOR.json
  const stateVectorPath = join(cwd, stateVectorRel)
  if (!existsSync(stateVectorPath)) {
    console.error(chalk.red(`  ✗ sync failed: STATE_VECTOR.json not found at ${stateVectorRel}`))
    console.error(chalk.gray('    → Run: workbench init to generate it'))
    process.exit(1)
  }

  const spinner = ora('Validating STATE_VECTOR.json...').start()

  let stateVector
  try {
    stateVector = JSON.parse(readFileSync(stateVectorPath, 'utf8'))
  } catch (err) {
    spinner.fail('Validation failed')
    console.error(chalk.red(`  ✗ parse STATE_VECTOR.json failed: invalid JSON`))
    console.error(chalk.gray(`    → Fix syntax errors in ${stateVectorRel} and retry`))
    process.exit(1)
  }

  const schema = loadSchema()
  if (schema) {
    const ajv = new Ajv({ allErrors: true })
    addFormats(ajv)
    const validate = ajv.compile(schema)
    if (!validate(stateVector)) {
      spinner.fail('Validation failed')
      const errors = validate.errors.map(e => `${e.instancePath || '/'}: ${e.message}`).join(', ')
      console.error(chalk.red(`  ✗ validate STATE_VECTOR.json failed: ${errors}`))
      console.error(chalk.gray(`    → Fix the listed fields in ${stateVectorRel} and retry`))
      process.exit(1)
    }
  }
  spinner.succeed('STATE_VECTOR.json is valid')

  // Dry run
  if (options.dryRun) {
    console.log(chalk.yellow('\n  Dry run — would sync:'))
    console.log(chalk.white(`    ${stateVectorRel} → projects/${projectName}/STATE_VECTOR.json`))
    if (existsSync(join(cwd, handoffRel))) {
      console.log(chalk.white(`    ${handoffRel} → projects/${projectName}/HANDOFF.md`))
    }
    console.log()
    return
  }

  // Resolve GitHub token
  const token = await resolveToken()
  const authUrl = buildAuthUrl(bridgeRepo, token)

  // Clone bridge repo
  const cloneSpinner = ora(`Cloning ${bridgeRepo}...`).start()
  const tmpDir = mkdtempSync(join(tmpdir(), 'workbench-'))

  try {
    const git = simpleGit()
    try {
      await git.clone(authUrl, tmpDir, ['--depth', '1'])
    } catch (err) {
      cloneSpinner.fail('Clone failed')
      printGitError('clone', err)
      process.exit(1)
    }
    cloneSpinner.succeed(`Cloned ${bridgeRepo}`)

    // Copy state files into bridge repo
    const targetDir = join(tmpDir, 'projects', projectName)
    try {
      mkdirSync(targetDir, { recursive: true })
    } catch (err) {
      console.error(chalk.red(`  ✗ create project directory failed: ${err.message}`))
      console.error(chalk.gray(`    → Check disk space and permissions on temp directory`))
      process.exit(1)
    }

    copyFileSync(stateVectorPath, join(targetDir, 'STATE_VECTOR.json'))
    console.log(chalk.green('  Copied: STATE_VECTOR.json'))

    const handoffPath = join(cwd, handoffRel)
    if (existsSync(handoffPath)) {
      copyFileSync(handoffPath, join(targetDir, 'HANDOFF.md'))
      console.log(chalk.green('  Copied: HANDOFF.md'))
    }

    const nextTaskPath = join(cwd, 'docs', 'NEXT_TASK.md')
    if (existsSync(nextTaskPath)) {
      copyFileSync(nextTaskPath, join(targetDir, 'NEXT_TASK.md'))
      console.log(chalk.green('  Copied: NEXT_TASK.md'))
    }

    // Commit
    const bridgeGit = simpleGit(tmpDir)
    await bridgeGit.addConfig('user.name', 'workbench-ai')
    await bridgeGit.addConfig('user.email', 'workbench-ai@users.noreply.github.com')

    try {
      await bridgeGit.add('-A')
    } catch (err) {
      printGitError('add', err)
      process.exit(1)
    }

    const status = await bridgeGit.status()
    if (status.files.length === 0) {
      console.log(chalk.yellow('\n  No changes to push — bridge repo is already up to date.\n'))
      return
    }

    const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 16) + ' UTC'
    const commitMsg = `workbench sync: ${projectName} -- ${timestamp}`

    try {
      await bridgeGit.commit(commitMsg)
    } catch (err) {
      printGitError('commit', err)
      process.exit(1)
    }

    // Push
    const pushSpinner = ora('Pushing to bridge repo...').start()
    try {
      await bridgeGit.push('origin', 'main')
    } catch (err) {
      pushSpinner.fail('Push failed')
      printGitError('push', err)
      process.exit(1)
    }

    const log = await bridgeGit.log({ maxCount: 1 })
    const sha = log.latest.hash.substring(0, 7)

    pushSpinner.succeed(`Pushed: ${chalk.bold(sha)}`)
    console.log(chalk.green(`\n  Sync complete for ${chalk.bold(projectName)}`))
    console.log(chalk.gray(`  ${commitMsg}\n`))
  } finally {
    try { rmSync(tmpDir, { recursive: true, force: true }) } catch {}
  }
}

File: packages/python/contextkeeper/sync.py
"""contextkeeper sync — push state files to the bridge repo."""

from __future__ import annotations

import json
import shutil
import tempfile
from datetime import datetime, timezone
from pathlib import Path

import jsonschema
from git import Repo
from rich.console import Console
from rich.spinner import Spinner

console = Console()


def _load_config(cwd: Path) -> dict | None:
    config_path = cwd / ".workbench"
    if not config_path.exists():
        return None
    return json.loads(config_path.read_text(encoding="utf-8"))


def _load_schema() -> dict | None:
    # Try relative to this package (installed from repo)
    candidates = [
        Path(__file__).resolve().parent.parent.parent.parent / "protocol" / "workbench.schema.json",
        Path.home() / ".workbench" / "src" / "protocol" / "workbench.schema.json",
    ]
    for path in candidates:
        if path.exists():
            return json.loads(path.read_text(encoding="utf-8"))
    return None


def sync_project(bridge: str | None = None, dry_run: bool = False) -> None:
    """Sync state files to the workbench bridge repo."""
    cwd = Path.cwd()
    config = _load_config(cwd)

    console.print("\n  [cyan]contextkeeper sync[/cyan]\n")

    if not config and not bridge:
        console.print("  [red]No .workbench config found. Run contextkeeper init first, or pass --bridge.[/red]")
        raise SystemExit(1)

    bridge_repo = bridge or (config and config.get("bridge_repo"))
    project_name = config.get("project_name", "") if config else ""
    sv_rel = config.get("state_vector_path", "handoff/STATE_VECTOR.json") if config else "handoff/STATE_VECTOR.json"
    handoff_rel = config.get("handoff_path", "docs/HANDOFF.md") if config else "docs/HANDOFF.md"

    if not bridge_repo:
        console.print("  [red]No bridge repo configured. Run contextkeeper init or pass --bridge.[/red]")
        raise SystemExit(1)

    if not project_name:
        console.print("  [red]No project_name in .workbench config.[/red]")
        raise SystemExit(1)

    # Read and validate STATE_VECTOR.json
    sv_path = cwd / sv_rel
    if not sv_path.exists():
        console.print(f"  [red]STATE_VECTOR.json not found at {sv_path}[/red]")
        raise SystemExit(1)

    with console.status("Validating STATE_VECTOR.json..."):
        state_vector = json.loads(sv_path.read_text(encoding="utf-8"))
        schema = _load_schema()
        if schema:
            try:
                jsonschema.validate(instance=state_vector, schema=schema)
            except jsonschema.ValidationError as e:
                console.print(f"  [red]Validation failed: {e.message}[/red]")
                raise SystemExit(1)

    console.print("  [green]STATE_VECTOR.json is valid[/green]")

    if dry_run:
        console.print("\n  [yellow]Dry run — would sync:[/yellow]")
        console.print(f"    {sv_rel} → projects/{project_name}/STATE_VECTOR.json")
        handoff_path = cwd / handoff_rel
        if handoff_path.exists():
            console.print(f"    {handoff_rel} → projects/{project_name}/HANDOFF.md")
        console.print()
        return

    # Clone bridge repo, copy files, commit, push
    tmp_dir = Path(tempfile.mkdtemp(prefix="workbench-"))
    try:
        bridge_url = f"https://github.com/{bridge_repo}.git"

        with console.status("Cloning bridge repo..."):
            repo = Repo.clone_from(bridge_url, str(tmp_dir), depth=1)

        console.print("  [green]Bridge repo cloned[/green]")

        target_dir = tmp_dir / "projects" / project_name
        target_dir.mkdir(parents=True, exist_ok=True)

        shutil.copy2(str(sv_path), str(target_dir / "STATE_VECTOR.json"))
        console.print("  [green]Copied: STATE_VECTOR.json[/green]")

        handoff_path = cwd / handoff_rel
        if handoff_path.exists():
            shutil.copy2(str(handoff_path), str(target_dir / "HANDOFF.md"))
            console.print("  [green]Copied: HANDOFF.md[/green]")

        next_task_path = cwd / "docs" / "NEXT_TASK.md"
        if next_task_path.exists():
            shutil.copy2(str(next_task_path), str(target_dir / "NEXT_TASK.md"))
            console.print("  [green]Copied: NEXT_TASK.md[/green]")

        # Commit and push
        with console.status("Pushing to bridge repo..."):
            repo.index.add("*")
            if not repo.is_dirty(untracked_files=True):
                console.print("  [yellow]No changes to push[/yellow]")
                return

            timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M UTC")
            commit_msg = f"chore(workbench): sync {project_name} -- {timestamp}"
            repo.index.commit(commit_msg)
            repo.remotes.origin.push()

        sha = repo.head.commit.hexsha[:7]
        console.print(f"  [green]Pushed: [bold]{sha}[/bold][/green]")
        console.print(f"\n  [green]Sync complete for [bold]{project_name}[/bold][/green]")
        console.print(f"  [dim]{commit_msg}[/dim]\n")

    finally:
        shutil.rmtree(tmp_dir, ignore_errors=True)
